### ISSUE TITLE
Pixelation of Vectors

### ISSUE THREAD

**Issue by SamratSK** [stephanrauh/ngx-extended-pdf-viewer#1095]
Hey Stephen,
I found this bug while testing a pdf. As I zoom in, the vectors like lines and text which should not pixelate ideally(and remain sharp and crisp) are pixelating at the edges. I understand that this is because the app is rendering the pdf on a canvas, which basically makes it an image. Is there a way to overcome this by using some option during Pdf.js rendering or using WebGL?

I tried setting the I have tried setting the pdfDefaultOptions.enableWebGL to true without any luck. Do you know any way to overcome it?

Here is a screenshot of the pixelation.

![Pixellation](https://user-images.githubusercontent.com/84336865/146524505-ff5da5c6-e3c8-4caf-a00e-19b7e0e01c3f.png)

Thank You!

**Comment by SamratSK** [comment]
Please check this stackoverflow question: https://stackoverflow.com/questions/40066166/canvas-text-rendering-blurry and also this link: http://wdobbie.com/pdf/

Thank You

**Comment by stephanrauh** [comment]
@timvandermeij Can you give us a hint how to improve the rendering quality when zooming in? It's possible I've introduced a bug in my fork of pdf.js, but what does the base library offer?

@SamratSK I wonder if your lines and texts are really vector graphics. Can you double-check that? When I zoom the default demo of the showcase to 1000%, the text still renders sharp:

![image](https://user-images.githubusercontent.com/3045767/147000090-a68cf39b-18f3-4710-ae13-a68ee04009ca.png)

BTW, if you zoom using a pinch gesture, it's possible the text is blurry in the short period of time of the pinch gesture. But I suppose you've not talking about this.

Finally, please apologize my late answer. For some reason, your ticket slipped my attention.

**Comment by timvandermeij** [comment]
It has been a while since I last looked at this, but I think I gathered all information again. It's perhaps a bit complex, so if it's not clear, feel free to let me know.

PDF.js renders the pages onto a canvas, so it's indeed comparable to a raster image and not a vector image[1]. The fact that it's a raster image means that zooming in far will make rendering pixelated (obviously depending on the resolution/quality with which the canvas is rendered).

This provides us with a trade-off: on the one hand you want the highest quality image, but on the other hand you don't want to consume a lot more memory (and the higher the image quality, the more memory usage is required for it). Therefore, PDF.js uses a heuristic during zooming to determine if it needs to re-render the page at a higher resolution, _or_ (and that's what you're seeing) do CSS zooming. The latter means that we don't re-render the canvas at a higher resolution but we simply zoom the existing canvas using CSS. This obviously leads to inferior quality and therefore the heuristic is there to limit doing that until it's absolutely necessary, i.e., until the point where memory usage would simply be too large for browsers to reasonably handle.

You can find the implementation at https://github.com/mozilla/pdf.js/blob/869b396011b03d00325f2a5d5c01e58841efc312/web/pdf_page_view.js#L378-L414, but it boils down to calculating how large the resulting canvas will be if the zoom factor is applied and checking that it doesn't go over a configurable threshold. If it's under the threshold, we'll re-render the canvas at a higher resolution so you get vector-like quality. If it's over the threshold, we can only fall back to CSS zooming. I believe in the past we always did CSS zooming if the zoom factor exceeded 400%, but it got smarter over time and now we're doing this calculation so also higher zoom factors will still look as sharp as possible.

In short (well, "short" is perhaps not the right word for this wall of text...), that's what the default viewer in PDF.js does. To the question "what can we do about it?", there is really only one answer: you can try increasing the threshold for the canvas back-end, which we made configurable via the render parameters (see https://github.com/mozilla/pdf.js/blob/869b396011b03d00325f2a5d5c01e58841efc312/web/pdf_page_view.js#L82-L84), to force the viewer to favor re-rendering over CSS zooming more often. However, this does come at the price of (potentially much) more memory usage, so that requires very careful consideration and benchmarking. Keep in mind that the current value is there for a good reason.

[1] PDF.js does have an _experimental_ SVG back-end to render the pages as true vector images, but mainly due to missing features and performance problems in the SVG rendering engines of browsers that has not become a suitable replacement so far, so we explicitly _don't_ recommend/support it.

**Comment by stephanrauh** [comment]
@timvandermeij Maybe you're interested to learn that I've tweaked this algorithm some time ago: https://github.com/stephanrauh/pdf.js/blob/bleeding-edge/web/pdf_page_view.js#L842-L862

At the time, the idea was to ensure that mobile browsers cope with the canvas size. If you don't care about memory and performance, there's a lot headroom beyond the 16 megapixel limit pdf.js imposes.

It's interesting I didn't see the 16 megapixel limit just a few lines above. I suppose I should re-work the source code again.

**Comment by stephanrauh** [comment]
@SamratSK In the meantime, you can lift the 16 megapixel restriction by setting `pdfDefaultOptions.maxCanvasPixels = -1`.

**Comment by SamratSK** [comment]
Hi @stephanrauh,
First of all, my apologies for the late message.

For your first message concerning whether the PDF is vector, I checked it and it is vector.

About setting pdfDefaultOptions.maxCanvasPixels = -1, sure will do that and let you know soon.

Thanks.

**Comment by stephanrauh** [comment]
I've improved the rendering algorithm. Now it activates CSS zooming if the requested zoom exceed the capabilities of the browser. I've tested it with a magnification of up to 25, and the PDF looks remarkable crisp. The example doesn't need CSS magnification until 10x magnification, which is even sharper. You see the difference immediately when you set `pdfDefaultOptions.maxCanvasPixels = -1`.

![image](https://user-images.githubusercontent.com/3045767/147242392-011e77eb-3842-40d0-8953-93409ca28461.png)

![image](https://user-images.githubusercontent.com/3045767/147244459-c3d866af-1703-4d59-a8d0-609fd36e0d10.png)
